docs/0.3/workflow_specification.md

```diff
@@ -0,0 +1,182 @@
+        for state in &self.states {
+            if state.name() == name {
+                return Some(state.clone());
+            }
+        }
+
+        None
+    }
+
+    pub fn starting_states(&self) -> &[String] {
+        &self.starting_states
+    }
+}
+```
+
+### WorkflowState
+
+`WorkflowState` has a name, list of constraints, list of permission aliases,
+and list of transitions that can be made from that state. It has two methods:
+`can_transition` and `expand_permission`. `can_transition` returns true if an
+entity can execute a transition to a given state given its Pike permissions.
+`expand_permissions` returns a list of all permissions that are stored under a
+given `PermissionAlias`:
+
+```rust
+#[derive(Clone)]
+pub struct WorkflowState {
+    name: String,
+    constraints: Vec<String>,
+    permission_aliases: Vec<PermissionAlias>,
+    transitions: Vec<String>,
+}
+
+impl WorkflowState {
+    pub fn can_transition(&self, new_state: String, pike_permissions: &[String]) -> bool {
+        if self.name == new_state {
+            return true;
+        }
+
+        if !self.transitions.contains(&new_state) {
+            return false;
+        }
+
+        for perm in pike_permissions {
+            for alias in &self.permission_aliases {
+                if alias.name() == perm && alias.transitions.contains(&new_state) {
+                    return true;
+                }
+            }
+        }
+
+        false
+    }
+
+    pub fn expand_permissions(&self, names: &[String]) -> Vec<String> {
+        let mut perms = Vec::new();
+
+        for name in names {
+            for alias in &self.permission_aliases {
+                if alias.name() == name {
+                    perms.append(&mut alias.permissions().to_vec());
+                }
+            }
+        }
+
+        perms
+    }
+}
+```
+
+### PermissionAlias
+
+`PermissionAlias` is an alias that houses multiple permissions. It has a name, a
+list of actions that live under the alias, and a list of transitions the alias
+can perform:
+
+```rust
+#[derive(Clone, Default)]
+pub struct PermissionAlias {
+    name: String,
+    permissions: Vec<String>,
+    transitions: Vec<String>,
+}
+
+impl PermissionAlias {
+    pub fn new(name: &str) -> Self {
+        Self {
+            name: name.to_string(),
+            permissions: vec![],
+            transitions: vec![],
+        }
+    }
+
+    pub fn add_permission(&mut self, permission: &str) {
+        self.permissions.push(permission.to_string());
+    }
+
+    pub fn add_transition(&mut self, transition: &str) {
+        self.transitions.push(transition.to_string());
+    }
+}
+```
+
+Below is an example of a permission alias with the name `po::seller` that has
+permission to perform two actions, update a purchase order and update its
+version number, and the ability to perform one transition, move the purchase
+order to it’s confirmed state:
+
+```rust
+let alias = PermissionAlias::new("po::seller");
```

Comment:
It looks like there is a `}` at the end of this code example that I don't think should be there

Reply:
Also, the permission name should match the thing being assigned in these lines below, so the following should be changed to `alias.add_permission(...)`. And, the indentation of these lines is not needed.

Reply:
Thanks, updated this with a real permissions example.